tensorflow/core/kernels/mkl_tmp_bf16_ops.cc

```diff
@@ -0,0 +1,62 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// See docs in ../ops/nn_ops.cc.
+
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/kernels/no_op.h"
+
+#ifdef ENABLE_INTEL_MKL_BFLOAT16
+
+namespace tensorflow {
+
+// This file contains temporary registrations for some of the Eigen CPU backend
+// operators for BFloat16 type. The kernel registered for all these ops is a
+// NoOp. We do this so that MKL graph pass can rewrite these ops into
+// corresponding MKL ops. Without such registrations, Placer component in
+// TensorFlow fails because Eigen CPU backend does not support these ops in
+// BFloat16 type.
+
+#define REGISTER_CPU(T)                                                       \
+  REGISTER_KERNEL_BUILDER(                                                    \
+      Name("Conv2D").Device(DEVICE_CPU).TypeConstraint<T>("T"), NoOp);        \
+  REGISTER_KERNEL_BUILDER(                                                    \
+      Name("Conv2DBackpropFilter").Device(DEVICE_CPU).TypeConstraint<T>("T"), \
+      NoOp);                                                                  \
+  REGISTER_KERNEL_BUILDER(                                                    \
+      Name("Conv2DBackpropInput").Device(DEVICE_CPU).TypeConstraint<T>("T"),  \
+      NoOp);                                                                  \
+  REGISTER_KERNEL_BUILDER(                                                    \
+      Name("AvgPool").Device(DEVICE_CPU).TypeConstraint<T>("T"), NoOp);       \
+  REGISTER_KERNEL_BUILDER(                                                    \
+      Name("AvgPoolGrad").Device(DEVICE_CPU).TypeConstraint<T>("T"), NoOp);   \
+  REGISTER_KERNEL_BUILDER(Name("FusedBatchNormV3")                            \
+                              .Device(DEVICE_CPU)                             \
+                              .TypeConstraint<bfloat16>("T")                  \
+                              .TypeConstraint<float>("U"),                    \
+                          NoOp);                                              \
+  REGISTER_KERNEL_BUILDER(Name("FusedBatchNormGradV3")                        \
+                              .Device(DEVICE_CPU)                             \
+                              .TypeConstraint<bfloat16>("T")                  \
+                              .TypeConstraint<float>("U"),                    \
+                          NoOp);
+
+TF_CALL_bfloat16(REGISTER_CPU);
```

Comment:
Note that these registrations cause TF to think we actually DO have a bf16 Conv2D (etc...) on CPU, and tries to execute it since there is no Eigen version to override it.  For example,
```
import tensorflow as tf
dtype = tf.bfloat16
x = tf.constant(1, dtype=dtype, shape=(3, 3, 3, 3))
kernel = tf.constant(1, dtype=dtype, shape=(1, 1, 1, 3))
z = tf.nn.conv2d(x, kernel, strides=[1, 1, 1, 1], padding='VALID')
print(z)
```
Returns the empty tensor:
```
tf.Tensor([], shape=(0,), dtype=float32)
```
This is causing crashes further down-the-line.

This even fails in graph mode:
```
@tf.function
def conv2d(x, kernel):
  return tf.nn.conv2d(x, kernel, strides=[1, 1, 1, 1], padding='VALID')

z = conv2d(x, kernel)
```
fails with
```
InternalError:  Missing 0-th output from node Conv2D (defined at <ipython-input-12-9e046227cc36>:3)  [Op:__inference_conv2d_70]
```

Reply:
cc: @agramesh1 @TensorFlow-MKL

Reply:
hi, `kernel = tf.constant(1, dtype=dtype, shape=(1, 1, 1, 3)) should be  tf.constant(1, dtype=dtype, shape=(1, 1, 3, 3)`
```
import tensorflow as tf
x = tf.constant(1, dtype=tf.bfloat16, shape=(3, 3, 3, 3))
kernel = tf.constant(1, dtype=tf.bfloat16, shape=(1, 1, 3, 3))
z = tf.nn.conv2d(x, kernel, strides=[1, 1, 1, 1], padding='VALID')
```
```
print(z)
<tf.Tensor: shape=(3, 3, 3, 3), dtype=bfloat16, numpy=
array([[[[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]],

        [[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]],

        [[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]]],


       [[[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]],

        [[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]],

        [[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]]],


       [[[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]],

        [[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]],

        [[3, 3, 3],
         [3, 3, 3],
         [3, 3, 3]]]], dtype=bfloat16)>
```

it is ok, what is your tf version? @cancan101 @penpornk  my tf version is latest master branch tf2.6

Reply:
Starts at 2.5, still fails with nightly:

https://colab.research.google.com/drive/1DXlD9u1yWr-b0VChJbGCPtw_G7kp6HKi?usp=sharing

Reply:
bazel build  --cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0 --copt=-mavx --copt=-mavx2 --copt=-mfma --copt=-mavx512f --copt=-mavx512pf --copt=-mavx512cd --copt=-mavx512dq  --config=opt -s -c opt //tensorflow/tools/pip_package:build_pip_package


you could build tensorflow from source.  use --config=mkl,

Reply:
@cantonios

Reply:
@gyshi I'm not trying to run these kernels.  I'm flagging that these are currently breaking real-world models in the wild, ever since OSS TF started building with mkl by default.  They should not exist in the pre-built pip packages.

These NoOp registrations need to be either protected so they are built conditionally, or removed and replaced with functioning default kernels.  As is, these NoOps are sometimes conditionally called, returning empty tensors (e.g. when trying to enable mixed precision), resulting in crashes and difficult-to-trace error messages.  We can't ask all users to "build the pip package from scratch if your model happens to crash because it's trying to run one of MKL's registered NoOps".